### PR TITLE
54 implement room participation feature

### LINKED ISSUE #54: Implement Room Participation Feature
## Objectives
- Create a seamless room joining experience for users
- Implement secure validation of room codes
- Enable real-time updates of participant status changes
- Establish reliable WebSocket connections for session participation
- Prevent username conflicts within rooms

## Challenges
- Ensuring WebSocket reliability across different network conditions
- Handling race conditions when multiple users join simultaneously
- Managing reconnection scenarios without creating duplicate participants
- Optimizing performance for rooms with many participants
- Securing the room access to prevent unauthorized entry

## Completion Criteria
- Users can join rooms using valid room codes
- Duplicate usernames are prevented within the same room
- All room participants receive real-time notifications when users join
- Online status of participants is accurately displayed and updated
- The system gracefully handles disconnections and reconnections
- All API endpoints return appropriate status codes and error messages

## Steps
1. Develop Room Code Validation API

- [x]    - Create endpoint to verify room code existence and validity
- [x]    - Implement response with room status information

2. Build Session Information Retrieval

- [x]    - Create endpoint to fetch session details based on room code
- [x]    - Include session configuration and status in response

3. Implement Username Duplication Check

- [x]    - Create endpoint to verify username availability in the room
- [x]    - Return appropriate error messages for taken usernames

4. Develop Participant Registration API

- [x]    - Create endpoint to register users to rooms
- [x]    - Store participant data in the database

5. Set up WebSocket Infrastructure

- [x]    - Configure WebSocket server for session rooms
- [x]    - Implement connection handling logic

6. Implement Socket Room Joining

- [x]    - Create logic to add users to socket rooms
- [x]    - Handle authentication and validation

7. Develop Online Status Management

- [x]    - Create system to track and update participant status
- [x]    - Implement reconnection handling

8. Add Join Completion Notification

- [x]    - Send confirmation to newly joined users
- [x]    - Include necessary session information

9. Implement All-Participant Updates

- [x]    - Broadcast join events to existing participants
- [x]    - Include new participant information in broadcasts

10. Test and Debug

- [x]     - Verify all endpoints and WebSocket functionality
- [x]     - Test edge cases and error handling

### Room Code Validation Request
```json
{
  "roomCode": "ABC123"
}
```

### Room Code Validation Response
```json
{
  "valid": true,
  "roomStatus": "active",
  "roomId": "room_12345"
}
```

### Session Information Response
```json
{
  "sessionId": "session_12345",
  "name": "Quiz Session",
  "status": "in_progress",
  "participantCount": 12,
  "configuration": {
    "maxParticipants": 50,
    "allowLateJoin": true
  }
}
```

### Username Check Request
```json
{
  "roomId": "room_12345",
  "username": "User123"
}
```

### Username Check Response
```json
{
  "available": true
}
```

### Participant Registration Request
```json
{
  "roomId": "room_12345",
  "username": "User123",
  "userId": "user_789"
}
```

### WebSocket Connection Structure
```json
{
  "type": "join_room",
  "roomId": "room_12345",
  "participant": {
    "id": "participant_456",
    "username": "User123",
    "joinedAt": "2025-04-16T14:30:00Z"
  }
}
```

### Participant Join Notification
```json
{
  "type": "participant_joined",
  "roomId": "room_12345",
  "participant": {
    "id": "participant_456",
    "username": "User123",
    "status": "online"
  },
  "participantCount": 13
}
```

### PR DESCRIPTION
**Pull request by kazuki-0418**
## Summary

This Pull Request implements the core functionality for users to join rooms, focusing on a seamless and secure experience. It includes API endpoints for room code validation, session information retrieval, and participant registration, as well as WebSocket integration for real-time updates. Key features include preventing duplicate usernames, handling participant status changes, and ensuring WebSocket reliability.

## Linked Issue

Closes #54 

## Changes

This PR introduces the following changes to achieve the room joining objectives:

* **API Endpoints:**
    * `/api/rooms/validate/:roomCode`:  Endpoint to verify room code existence and validity. Returns `valid`, `roomStatus`, and `roomId`. (Implemented in Step 1)

* **WebSocket Integration:**

    * WebSocket server configured for session rooms.
    * Logic for adding users to socket rooms, including authentication and validation of the `join_room` request. (Implemented in Step 6)
    * System to track and update participant online status, including reconnection handling. (Implemented in Step 7)
    * The following WebSocket events are implemented:
        * `SESSION_JOIN_REQUEST`: Request for a user to join a session.
        * `SESSION_LEAVE_REQUEST`: Request for a user to leave a session.
        * `SESSION_JOIN_SUCCESS`: Notification indicating successful session joining.
        * `SESSION_LEAVE_SUCCESS`: Notification indicating successful session leaving.
        * `SESSION_DATA`: Event to request session-related data.
        * `SESSION_DATA_RESPONSE`: Event to send session data in response to a `SESSION_DATA` event.
        * `SESSION_CLOSE_REQUEST`: Request to close a session.
        * `SESSION_CLOSE_SUCCESS`: Notification indicating successful session closure.
        * `PARTICIPANT_JOINED`: Notification indicating a new participant has joined the session.
        * `PARTICIPANT_LEFT`: Notification indicating a participant has left the session.
        * `PARTICIPANT_UPDATE`: Notification indicating participant information has been updated (e.g., status change).
        * `PARTICIPANT_KICKED`: Notification indicating a participant has been forcibly removed from the session.
        * `join_room`: Initial request to join a room. (Consider retaining or removing this older event name)
        * `participant_joined`: Notification indicating a new participant has joined. (Consider retaining or removing this older event name)

* **Real-time Updates:**

    * New participants receive a confirmation with session information via the `SESSION_JOIN_SUCCESS` event. (Implemented in Step 8)
    * The `PARTICIPANT_JOINED` event is broadcast to existing participants to notify them of a new user joining the room, including the new participant's details and the updated participant count. (Implemented in Step 9)
    * Participant online status is updated in real-time via the `PARTICIPANT_UPDATE` event.
    * The `PARTICIPANT_LEFT` event is sent when a user leaves the session.
    * The `SESSION_CLOSE_SUCCESS` event is sent when a session is closed.

* **Validation and Error Handling:**
    * Room code validation is implemented to prevent invalid room entries.
    * Username duplication is prevented, and appropriate error messages are returned.
    * WebSocket connection handling includes authentication and validation to secure room access.
    * API endpoints return appropriate HTTP status codes to indicate success or failure.

* **Testing and Debugging:**
    * All API endpoints and WebSocket functionality have been verified. (Implemented in Step 10)
    * Edge cases and error handling have been tested (e.g., invalid room codes, duplicate usernames, network disconnections).

**NOTE:** This PR lays the foundation for room joining. Further PRs may be needed to address specific UI/UX aspects, performance optimizations under extreme load, and more robust error handling for all potential failure scenarios.

## Verification Results

The following verification steps were performed:

* **API Endpoint Testing:**
    * Used tools like Postman/Insomnia to manually test all API endpoints:
        * Verified successful responses for valid inputs.
        * Confirmed appropriate error status codes (e.g., 400 Bad Request, 404 Not Found) for invalid inputs (e.g., non-existent room code, duplicate username).
    * Included unit tests (if applicable - specify framework used) to automate testing of API logic.

* **WebSocket Testing:**
    * Used a WebSocket client (e.g., Simple WebSocket Client) to simulate multiple users joining and leaving rooms.
    * Verified that:
        * Users can successfully join rooms with valid room codes.
        * `participant_joined` messages are broadcast to all connected clients in the room.
        * The `participantCount` in the `participant_joined` message is accurate.
        * Reconnection scenarios are handled correctly (no duplicate participants).
        * Error messages are sent for invalid join requests (e.g., invalid room code, duplicate username).

* **Database Verification:**
    * Manually inspected the database to confirm that participant data is stored correctly.
    * Verified that participant status updates are reflected in the database (if applicable).

* **Screenshots:**
    * (Attach screenshots of Postman/Insomnia tests showing successful API calls and error responses).
    * (Attach screenshots of the WebSocket client showing successful room joins and `participant_joined` messages).
    * (Attach screenshots of the database showing participant data).

    *(Example: "Screenshot 1: Postman - Successful room code validation")*
    
image

    *(Example: "Screenshot 2: WebSocket Client - User A joining room, User B receiving notification")*

https://github.com/user-attachments/assets/dbb73e5b-2eaa-4f3d-b932-d51ec4560568

**Important Considerations for Reviewers:**

* **Error Handling:** Pay close attention to how edge cases and potential errors are handled (e.g., network failures, database connection issues).
* **Security:** Review the authentication and validation logic in the WebSocket connection handling to ensure secure room access.
* **Performance:** Consider potential performance bottlenecks, especially with a large number of concurrent users.
* **Scalability:** Think abo

### CODE REVIEW

File: frontend/src/app/(with-websocket)/rooms/[roomCode]/page.tsx [hunk 1]
```diff
@@ -2,11 +2,10 @@
 import { UserAccessOptions } from "@/components/pages/rooms/UserAccessOptions";
 import { FullHeightCardLayout } from "@/components/ui/FullHeightCardLayout";
 
-export default function RoomsHostPage() {
-  const allowGuests = true; // TODO: Replace with actual logic to check if the user is a guest
+export default function RoomsUserPage() {
   return (
     <FullHeightCardLayout useWithHeader={false}>
-      <UserAccessOptions allowGuests={allowGuests} />
+      <UserAccessOptions allowGuests={true} />
```
Comment (ayakakojimaak) [inline frontend/src/app/(with-websocket)/rooms/[roomCode]/page.tsx:8]: To confirm, it looks like we allow guests at this point, is that right?

File: frontend/src/app/(with-websocket)/sessions/[sessionId]/page.tsx [hunk 2]
```diff
@@ -0,0 +1,228 @@
+"use client";
```
Comment (ayakakojimaak) [inline frontend/src/app/(with-websocket)/sessions/[sessionId]/page.tsx]: For the future, the plan is to tidy up the content in `frontend/src/app/sessions/[sessionId]/page.tsx` and move it over here.
Does that sound like I've got it right? 

Sorry for any extra useState or props that might be in the page.tsx.
Please let me know if there's anything I can do to help with that!

**Review by ayakakojimaak — COMMENTED** [review]
You've made so many fixes and hooked up the backend.
Thank you so much for that!
Just sending over two quick things to double-check.

**Review by ayakakojimaak — APPROVED** [review]
I believe `main` is unrestricted, but I'll send the approval just in case.

LGTM

File: frontend/src/app/(with-websocket)/rooms/[roomCode]/page.tsx (hunk 1 above)
Comment (kazuki-0418) [reply frontend/src/app/(with-websocket)/rooms/[roomCode]/page.tsx:8]: Exactly, and for now it's only for the guests, I haven't implemented for the user that is already logined
